### PR TITLE
Extend test helper coverage; simplify CUB axis test

### PR DESCRIPTION
**Pull request by leofang**
Split from #3740. This PR does two things:
1. Cover more valid contiguous axes for the tests: now negative axes are also tested. 
2. Improve `_can_use_cub_block_reduction()`: the tuples `out_axis` and `reduce_axis` there are returned by `_get_axis()`, which makes the axes always sorted and positive, so we can simply join them as we do in `cupy.cuda.cub._preprocess_array()`. But, this also means that for its tests we need to convert the negative axes to positive ones. Thanks @grlee77 for suggestions!

### CODE REVIEW

**Comment by grlee77** [comment]
The two files aside from the CUB tests where this `for_contiguous_axes` decorator is used are:
`cupy_tests/math_tests/test_sumprod.py`
`cupy_tests/core_tests/test_ndarray_reduction.py`

Those both pass without modification when I tested locally.

**Comment by leofang** [comment]
Jenkins, test this please.

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/testing/helper.py [hunk 1]
```diff
@@ -1098,6 +1100,9 @@ def test_func(self, *args, **kw):
                 try:
                     kw[name] = a
                     impl(self, *args, **kw)
+                    a = tuple(i - ndim for i in a)
```
Comment (asi1024) [inline cupy/testing/helper.py:1103]: Conventionally we write as follows because building tuple from a list is faster than from a generator.
```suggestion
                    a = tuple([i - ndim for i in a])
```

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by asi1024** [comment]
The decorator `for_contiguous_axes` seems for a low-level checks, but negative axes are removed by normalizing at preprocessing. Why not manually convert axes as `axis = [a - ndim for a in axis]` in each test?

**Comment by leofang** [comment]
Hi @asi1024 there are many tests using this decorator and I hope to test them all. Like you said, in many cases it should be normalized during preprocessing, and adding negative axes to the decorator will ensure the normalization happens as expected.

**Comment by asi1024** [comment]
We don't want to mix parameters for interface-level checks in to that low-level checks.

**Comment by asi1024** [comment]
Do you mean that `for_contiguous_axes` is only for CUB tests?

**Comment by leofang** [comment]
Currently yes!

**Comment by asi1024** [comment]
@leofang Sorry for my late response.
If this PR is only for the CUB test, could you make a fix in the CUB test code?

**Comment by leofang** [comment]
Sorry for dropping the ball, @asi1024. What fix did you have in mind?

**Comment by asi1024** [comment]
Sorry for my late response.
Could you add `negative_axis` parameter to `testing.product` in `TestSimpleCubReductionKernelContiguity`?

**Comment by leofang** [comment]
Closing it as the main fix to `_can_use_cub_block_reduction` is in. The test improvement no longer seems important to me, but I am happy to reopen if there's a need.
